digdag-core/src/main/java/io/digdag/core/workflow/AttemptBuilder.java

```diff
@@ -110,14 +111,26 @@ public AttemptRequest buildFromStoredWorkflow(
         return monitors.build();
     }
 
-    private Config buildSessionParameters(Config overrideParams, Optional<Scheduler> sr, Instant sessionTime, ZoneId timeZone)
+    private Config buildSessionParameters(Config overrideParams, Optional<Scheduler> sr, Instant sessionTime, ZoneId timeZone, Optional<Instant> lastExecutedSessionTime)
     {
         Config params = overrideParams.deepCopy();
         if (sr.isPresent()) {
             Instant lastSessionTime = sr.get().lastScheduleTime(sessionTime).getTime();
             Instant nextSessionTime = sr.get().nextScheduleTime(sessionTime).getTime();
             params.set("last_session_time", formatSessionTime(lastSessionTime, timeZone));
             params.set("next_session_time", formatSessionTime(nextSessionTime, timeZone));
+            if (lastExecutedSessionTime.isPresent()) {
+                // TODO: Come up with a better names for these parameters?
+
+                // This is the last session that was processed. Normally this will be the immediately preceding session,
+                // but it can be an earlier session if some were skipped due to skip_on_overtime = true.
+                params.set("last_processed_session_time", formatSessionTime(lastExecutedSessionTime.get(), timeZone));
+
+                // This is the first session time slot that should be included if the user wants to process any preceding skipped sessions.
+                // Normally this is the same as the current session_time but can be an earlier session if some were skipped.
+                Instant firstUnprocessedSessionTime = sr.get().nextScheduleTime(lastExecutedSessionTime.get()).getTime();
+                params.set("first_unprocessed_session_time", formatSessionTime(firstUnprocessedSessionTime, timeZone));
```

Comment:
> My thinking was that first_unprocessed_session_time would be equal to e.g. start_time for the first scheduled session

I see. That's a very good point.

I found the document of Metric Insights: [Write a SQL Statement for a Dimensioned Metric ](http://help.metricinsights.com/m/Creating_Dimensioned_Metrics_and_Reports/l/104398-write-a-sql-statement-for-a-dimensioned-metric)

> By instructing the system to **Batch Load Data Going back no more than** 14 days, :last_measurement_time would be set to be at most 14 days before the latest 'last measurement time' across all Dimension values. In this example, this would be 14 days before 2012-10-02 which will result in :last_measurement_time being set to '2012-09-18'.

For initial execution, Metric Insights subtract at most N days from the first scheduled time to set `:last_measurement_time`. Actually this behavior sounds useful.

How about this idea?:

* **last_session_time**
  * session_time of the last schedule. It doesn't matter whether the last session actually existed, failed, or skipped.
* **last_executed_session_time**
  * similar to _last_session_time_ but it goes back to actually executed session_time at most the period of `assume_executed` option. If `assume_executed` option is not set and this is the first execution, the value becomes empty string.
  * (I changed the name from `last_processed_session_time` because "processed" sounds confusing with "successful" for me)
* **last_successful_session_time**
  * similar to _last_session_time_ but it goes back to successfully finished session_time at most the period of `assume_successful` option. If `assume_successful` option is not set and there're no successful execution before, the value becomes empty string.

With above parameters, we can cover following 3 use cases:

### sessions run in parallel (currently supported)

```sql
schedule:
  daily>: 09:00:00

+analyze:
  td>:
  query: |
    ...
    WHERE '${last_session_time}' < date AND date < '${session_time}'
```

### session depends on the last execution, skip on overtime, failed execution always needs manual fix

```sql
schedule:
  daily>: 09:00:00
  skip_on_overtime: true
  assume_executed: 14 days

+analyze:
  td>:
  query: |
    ...
    WHERE '${last_executed_session_time}' < date AND date < '${session_time}'
```

### session depends on the last execution, skip on overtime, failed execution makes the next execution running longer

```sql
schedule:
  daily>: 09:00:00
  skip_on_overtime: true
  assume_successful: 14 days

+analyze:
  td>:
  query: |
    ...
    WHERE '${last_successful_session_time}' < date AND date < '${session_time}'
```